rust/src/nasl/builtin/host/mod.rs

```diff
@@ -134,169 +131,76 @@ fn nasl_get_host_ip(
-        t_aux.push_str(target_ori);
-        t_aux.push(']');
-        target = t_aux;
-    }
-
-    // SocketAddr requires a socket, not only the IP addr.
-    target.push_str(":5000");
-    match target.to_socket_addrs() {
-        Ok(addr) => {
-            let v = addr.into_iter().filter(|x| x.is_ipv6()).collect::<Vec<_>>();
-            Ok(NaslValue::Boolean(!v.is_empty()))
-        }
-        Err(_) => Err(FunctionErrorKind::diagnostic_ret_null("address is Null")),
-    }
+    Ok(target.parse::<Ipv6Addr>().is_ok())
 }
 
 /// Compare if two hosts are the same.
 /// The first two unnamed arguments are string containing the host to compare
 /// If the named argument cmp_hostname is set to TRUE, the given hosts are resolved into their hostnames
-fn same_host(register: &Register, _: &Context) -> Result<NaslValue, FunctionErrorKind> {
-    let positional = register.positional();
-    if positional.len() != 2 {
-        return Err(FunctionErrorKind::diagnostic_ret_null(
-            "same_host needs two parameters!",
-        ));
-    }
-
-    let h1 = match &positional[0] {
-        NaslValue::String(x) => {
-            if let Some(h) = resolve(x.to_string())? {
-                h
-            } else {
-                return Err(FunctionErrorKind::diagnostic_ret_null(
-                    "Wrong parameter type",
-                ));
-            }
-        }
-        _ => {
-            return Err(FunctionErrorKind::diagnostic_ret_null(
-                "Wrong parameter type",
-            ));
-        }
-    };
-
-    let h2 = match &positional[1] {
-        NaslValue::String(x) => {
-            if let Some(h) = resolve(x.to_string())? {
-                h
-            } else {
-                return Err(FunctionErrorKind::diagnostic_ret_null(
-                    "Wrong parameter type",
-                ));
-            }
-        }
-        _ => {
-            return Err(FunctionErrorKind::diagnostic_ret_null(
-                "Wrong parameter type",
-            ));
-        }
-    };
-
-    let cmp_hostname = match register.named("cmp_hostname") {
-        Some(ContextType::Value(NaslValue::Boolean(x))) => *x,
-        _ => false,
-    };
+#[nasl_function(named(cmp_hostname))]
+fn same_host(
+    h1: &str,
+    h2: &str,
+    cmp_hostname: Option<bool>,
+) -> Result<NaslValue, FunctionErrorKind> {
+    let h1 = resolve(h1.to_string())?;
+    let h2 = resolve(h2.to_string())?;
 
-    let addr1: Vec<IpAddr> = h1.into_iter().map(|x| x.ip()).collect::<Vec<_>>();
-    let addr2: Vec<IpAddr> = h2.into_iter().map(|x| x.ip()).collect::<Vec<_>>();
+    let addr1: Vec<IpAddr> = h1
+        .into_iter()
+        .map(|x| x.parse().unwrap())
+        .collect::<Vec<_>>();
+    let addr2: Vec<IpAddr> = h2
+        .into_iter()
+        .map(|x| x.parse().unwrap())
+        .collect::<Vec<_>>();
 
     let hostnames1 = addr1
         .clone()
         .into_iter()
-        .map(|x| lookup_addr(&x))
+        .filter_map(|x| lookup_addr(&x).ok())
```

Comment:
Haven't tested, but you should be able to get rid of the clone by using `.iter()` instead. Something like this:

```suggestion rust 
    let hostnames1 = addr1
        .iter()
        .filter_map(|x| lookup_addr(x).ok())
```